# Global Super E-commerce project

Global Superstore is a global online retailer based in New York, boasting a broad product catalog and aiming to be a one-stop shop for its customers.  Global Superstore’s clientele, hailing from 165 different countries, can browse through an endless offering with more than 10,000 products. This large selection comprises three main categories: **office supplies** (e.g.,  staples), **furniture**  (e.g.,  chairs), and **technology** (e.g.,  smartphones).

The Dataset is available in [Kaggle](https://www.kaggle.com/datasets/shekpaul/global-superstore), comprised by a spreadsheet with information on the 51,291  **Orders** that were processed between 2011 and 2014. The dataset provides 24 attributes including `Ship Date`, `City`, `Category`, and `Order Priority`.

Through this notebook I will be performing a basic EDA (Exploratory Data Analysis) over the previously mentioned dataset, using mostly pandas, which is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language.

 <p>
<div align="left">
<a href="https://pandas.pydata.org/" target="_blank"> <img src="https://drive.google.com/uc?export=view&id=16zbVoGIZzVCxEytvNxdCwkhZ9lieg3bg" style="height: 15rem"/> </a>
</div>
 </p>

Using the dataset downloaded from Kaggle, we have built the following ERD (Entity Relationship Diagram), which is basically a graphical representation that depicts relationships among the 3 sets of data inside the source file.

The Diagram is available in [dbdocs.io](https://dbdiagram.io/d/globa_superstore-6554bb497d8bbd64653dc307), and it consist in 3 tables, `orders`, `returns` and `people`

  ![erd](https://drive.google.com/uc?export=view&id=16vY1Ik0fDwPdl80TBTdxBHRZxi2R5CGJ)

Later in the analysis I will be analyzing the fields inside those tables.

I have set publicly available the dataset used in this notebook in the following [link](https://docs.google.com/spreadsheets/d/16ucuV7_pZ1FLxHPi_Teee8aKbT5oPUuR/edit?usp=drive_link&ouid=115237034183493852475&rtpof=true&sd=true)


## Importing libraries and data manipulation

In [1]:
import pandas as pd
import openpyxl
import os

In [2]:
## mounting google drive folders if not mounted
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


The input and output folder are available in the Google Drive following this [link](https://drive.google.com/drive/folders/167rq3O4-Gb2SBT42_gxdkqrV0CbJkN1Q)

In [3]:
## Setting variables for the input data and output data folders
input_data = '/content/drive/MyDrive/assestments/global_superstore_analysis/input_data/'
output_data = '/content/drive/MyDrive/assestments/global_superstore_analysis/output_data/'

## Data Analysis using pandas

In [5]:
def read_and_parse(path:str,
                   file_name:str,
                   low_memory=False) -> pd.DataFrame:
    """Reads and parses data from a CSV file to a pandas dataframe

    :param path: Path where is the CSV file
    :param file_name: Name of the file to be read
    :return: pandas DataFrame
    """
    #Reading CSV file from path
    df = pd.read_csv(f"{path}/{file_name}.csv")

    return df

def extract_csv_files(df: str,
                      path: str,
                      file_name: str):
    """Extracts a CSV from a DataFrame

    :param df: Name of the pandas DataFrame
    :param path: Path to store the CSV file
    :param file_name: Name of the file to be read
    :return: None
    """
    df.to_csv(f'{path}{file_name}.csv', encoding='utf-8', index=False)


def parse_xl_file(path: str,
                  sheet_name: str,
                  file_name: str) -> pd.DataFrame:

    """Parses an Excel file and returns a DataFrame

    :param path: Path where is the CSV file
    :param sheet_name: Name of the sheet of the Excel file
    :param file_name: Name of the file to be read
    :return: pandas DataFrame
    """
    xl = pd.ExcelFile(f"{path}{file_name}.xlsx")
    df = xl.parse(sheet_name)
    #converting header names to lowercase
    df.columns = df.columns.str.lower()

    return df

In [6]:
## Storing the file name in a variable for further use in the data manipulation process
file_name = os.listdir(input_data)
file_name = file_name[0]
print(f'The Excel file that contains the data that we will be analyzing is {file_name}\nand is stored in the {str.split(input_data,"/")[6]} folder of this directory')

The Excel file that contains the data that we will be analyzing is GlobalSuperstore.xlsx
and is stored in the input_data folder of this directory


In [7]:
gss = pd.ExcelFile(f"{input_data}{file_name}")
print(f'The Excel file has 3 tabs that corresponds to the 3 tables that where stored in the database {gss.sheet_names},\nand they will be manipulated further using basic pandas functions.')

The Excel file has 3 tabs that corresponds to the 3 tables that where stored in the database ['Orders', 'Returns', 'People'],
and they will be manipulated further using basic pandas functions.


## Analizing Global Superstore data with Python

The orders table has 24 fields and 51,291 colummns, that describes the transactions processed between 2011 and 2014. We will be analizing here in this notebook the entire dataset and then exporting it to the output folder as CSV format.

In [8]:
gss_orders = parse_xl_file(input_data,'Orders','GlobalSuperstore')
gss_orders.head(3)

row id        order id order date  ship date     ship mode customer id  \
0   32298  CA-2012-124891 2012-07-31 2012-07-31      Same Day    RH-19495   
1   26341   IN-2013-77878 2013-02-05 2013-02-07  Second Class    JR-16210   
2   25330   IN-2013-71249 2013-10-17 2013-10-18   First Class    CR-12730   

   customer name    segment           city            state  ...  \
0    Rick Hansen   Consumer  New York City         New York  ...   
1  Justin Ritter  Corporate     Wollongong  New South Wales  ...   
2   Craig Reiter   Consumer       Brisbane       Queensland  ...   

        product id    category sub-category  \
0  TEC-AC-10003033  Technology  Accessories   
1  FUR-CH-10003950   Furniture       Chairs   
2  TEC-PH-10004664  Technology       Phones   

                                        product name     sales quantity  \
0  Plantronics CS510 - Over-the-Head monaural Wir...  2309.650        7   
1          Novimex Executive Leather Armchair, Black  3709.395        9   
2                  Nokia Smart Phone, with Caller ID  5175.171        9   

  discount    profit  shipping cost  order priority  
0      0.0  762.1845         933.57        Critical  
1      0.1 -288.7650         923.63        Critical  
2      0.1  919.9710         915.49          Medium  

[3 rows x 24 columns]

Using one of the python functions written above, I will set a more readable format to the columns in the header.

In [9]:
#replacing the blanks with _ separator in each column name
gss_orders.columns = gss_orders.columns.str.replace(" ", "_")

In [10]:
gss_orders.columns

Index(['row_id', 'order_id', 'order_date', 'ship_date', 'ship_mode',
       'customer_id', 'customer_name', 'segment', 'city', 'state', 'country',
       'postal_code', 'market', 'region', 'product_id', 'category',
       'sub-category', 'product_name', 'sales', 'quantity', 'discount',
       'profit', 'shipping_cost', 'order_priority'],
      dtype='object')

The `info()` built-in pandas function, shows the user some useful statistics about this dataset, most of the columns has the required data to perform the analysis.

In [11]:
gss_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   row_id          51290 non-null  int64         
 1   order_id        51290 non-null  object        
 2   order_date      51290 non-null  datetime64[ns]
 3   ship_date       51290 non-null  datetime64[ns]
 4   ship_mode       51290 non-null  object        
 5   customer_id     51290 non-null  object        
 6   customer_name   51290 non-null  object        
 7   segment         51290 non-null  object        
 8   city            51290 non-null  object        
 9   state           51290 non-null  object        
 10  country         51290 non-null  object        
 11  postal_code     9994 non-null   float64       
 12  market          51290 non-null  object        
 13  region          51290 non-null  object        
 14  product_id      51290 non-null  object        
 15  ca

In [12]:
gss_orders.describe()

row_id   postal_code         sales      quantity      discount  \
count  51290.00000   9994.000000  51290.000000  51290.000000  51290.000000   
mean   25645.50000  55190.379428    246.490581      3.476545      0.142908   
std    14806.29199  32063.693350    487.565361      2.278766      0.212280   
min        1.00000   1040.000000      0.444000      1.000000      0.000000   
25%    12823.25000  23223.000000     30.758625      2.000000      0.000000   
50%    25645.50000  56430.500000     85.053000      3.000000      0.000000   
75%    38467.75000  90008.000000    251.053200      5.000000      0.200000   
max    51290.00000  99301.000000  22638.480000     14.000000      0.850000   

             profit  shipping_cost  
count  51290.000000   51290.000000  
mean      28.610982      26.375818  
std      174.340972      57.296810  
min    -6599.978000       0.002000  
25%        0.000000       2.610000  
50%        9.240000       7.790000  
75%       36.810000      24.450000  
max     8399.976000     933.570000

### Grouping data

We can generate descriptive tables that groups data from the **Orders** table by using the `groupby` function from pandas.

In [13]:
grouped_data = gss_orders.groupby('sub-category').agg({
    'sales': 'sum',
    'profit': 'sum',
    'quantity': 'sum',
    'order_id': 'count'  # Assuming 'Order ID' is a unique identifier for orders
}).reset_index()

# Format 'Sales' and 'Profit' columns as currency
grouped_data['sales'] = grouped_data['sales'].map('${:,.2f}'.format)
grouped_data['profit'] = grouped_data['profit'].map('${:,.2f}'.format)

# Rename the 'Order ID' column to 'Number of Orders' for clarity
grouped_data.rename(columns={'order_id': 'Number of Orders'}, inplace=True)

grouped_data

sub-category          sales       profit  quantity  Number of Orders
0   Accessories    $749,237.02  $129,626.31     10946              3075
1    Appliances  $1,011,064.30  $141,680.59      6078              1755
2           Art    $372,091.97   $57,953.91     16301              4883
3       Binders    $461,911.51   $72,449.85     21429              6152
4     Bookcases  $1,466,572.24  $161,924.42      8310              2411
5        Chairs  $1,501,681.76  $140,396.27     12336              3434
6       Copiers  $1,509,436.27  $258,567.55      7454              2223
7     Envelopes    $170,904.30   $29,601.12      8380              2435
8     Fasteners     $83,242.32   $11,525.42      8390              2420
9   Furnishings    $385,578.26   $46,967.43     11225              3170
10       Labels     $73,404.03   $15,010.51      9322              2606
11     Machines    $779,060.07   $58,867.87      4906              1486
12        Paper    $244,291.72   $59,207.68     12822              3538
13       Phones  $1,706,824.14  $216,717.01     11870              3357
14      Storage  $1,127,085.86  $108,461.49     16917              5059
15     Supplies    $243,074.22   $22,583.26      8543              2425
16       Tables    $757,041.92  $-64,083.39      3083               861

In [14]:
## Reading Global superstore Returns dataset
gss_returns = parse_xl_file(input_data,'Returns','GlobalSuperstore')
gss_returns.head(3)

returned         order id market
0      Yes   MX-2013-168137  LATAM
1      Yes   US-2011-165316  LATAM
2      Yes  ES-2013-1525878     EU

In [15]:
#replacing the blanks with _ separator in each column name
gss_returns.columns = gss_returns.columns.str.replace(" ", "_")
gss_returns.columns

Index(['returned', 'order_id', 'market'], dtype='object')

In [16]:
## Reading Global superstore People dataset
gss_people = parse_xl_file(input_data,'People','GlobalSuperstore')
gss_people.head(3)

person   region
0   Anna Andreadi  Central
1     Chuck Magee    South
2  Kelly Williams     East

## Extracting data from pandas dataframes to output data folder as CSV format

In [ ]:
extract_csv_files(gss_orders,output_data,'global_superstore_orders')

In [ ]:
extract_csv_files(gss_returns,output_data,'global_superstore_returns')

In [ ]:
extract_csv_files(gss_people,output_data,'global_superstore_people')

## Tableau Data Analysis

[Tableau](https://www.tableau.com/why-tableau/what-is-tableau) is a visual analytics platform that is commonly used to transform the way data practitioners use the data to solve data related problems. Is also a market-leading choice for modern business intelligence Teams and one of the  world’s most popular data visualization software packages

## Discovering Tableau Public

[Tableau Public](https://www.tableau.com/products/public) Tableau Public is a free platform to explore, create and publicly share data visualizations online. With the largest repository of data visualizations in the world to learn from, Tableau Public makes developing data skills easy. Advance your career in analytics by learning from limitless data inspiration and creating an online portfolio of work.

Main features:

* Share publicly: A platform for public (not private) data. Published visualizations are available for anyone to see online.

* Free with few limits: Explore and contribute to the millions of visualizations free. Saving locally and refreshing data are limited.

* Fully hosted: Tableau Public vizzes can handle millions of viewers. All infrastructure is managed by us at no cost.


## Insights and findings

The Dashboard developed in Tableau Public is available in the following [link](https://public.tableau.com/views/GlobalSuperstore_17006630958070/Main?:language=en-US&:display_count=n&:origin=viz_share_link)

From the visualizations I have performed from the data, the following are the findings I have found:

1. **Categories and sub categories**. The three categories present in the dataset for Global Superstore, **Technology**, **Furniture** and **Office Supplies**, groups 37.5%, 32.5%, and 30% of the total Sales respectively and they are almost evenly distributed.



 <p>
<div align="center">
<img src="https://drive.google.com/uc?export=view&id=1Al__nH4RIVeEcny9tFLU0oIVCmFo1VI0" style="height: 3rem"/>
</div>
<div align="center">
<em>Sales per Category Pie Chart from Tableau Dashboard</em>
</div>
 </p>




**Technology** accounts for the highest Total Profits (almost half) at 45.2%. **Office** supplies profits are at 35.3% while **Furniture** is at 19.4% profit.




 <p>
<div align="center">
<img src="https://drive.google.com/uc?export=view&id=1AxA0WhOnUgB9PoeYH3Mp2X4SdfwqJjU7" style="height: 3rem"/>
</div>
<div align="center">
<em>Profit per Category Area Chart from Tableau Dashboard</em>
</div>
 </p>


One of the questions that raises when we have analyzed the first 2 charts and the rest of the data is, Why do furniture contribute to such low profits yet has high sales?

Furniture category is comprises by 4 subcategories (`Furnishes`, `Bookcases`, `Chairs`, and `Tables`) Only `Furnishes` (12.2%) has a profit margin above the average profit margin of sub-categories (11.6%). Bookcases has an 11% profit margin, and chairs 9.3%.

Tables sub-category is infact making a loss, as it records a negative -8.5% profit margin, and specifically in this sub-category is where Global Superstore is losing money.


 <p>
<div align="center">
<img src="https://drive.google.com/uc?export=view&id=1AxIlpg5mTl0aKcvPY6-Gox7w5SqwWjDH" style="height: 3rem"/>
</div>
<div align="center">
<em>Profit, Sales and Margin per Category, table from Tableau Dashboard</em>
</div>
 </p>



2. Global superstore has three main **Segments**, **Consumer**, **Corporate** and **Home Office**. The Consumer segment accounts for over 6.5 Million dollars in sales 51.5%. Corporate 30.3% while Home Office 18.3%.

 <p>
<div align="center">
<img src="https://drive.google.com/uc?export=view&id=1Ay7_vCbmA03kXS5Pvca64aNFl_bB58UD" style="height: 3rem"/>
</div>
<div align="center">
<em>Sales per Segment Horizontal Bar Chart from Tableau Dashboard</em>
</div>
 </p>



3. The largest markets in Gobal Superstore, are the Asia Pacific market at 4042.6M, Europe 3287.3M and The US and Canada market 2364.1K. Canada has the lowest sales but highest percent profit margins among the markets. Global stores has a presence in major markets with profit margins mainly ranging between 10 to 14%. EMEA market has an uncharacteristic low profits and lower profit margin than the rest of the markets.



 <p>
<div align="center">
<img src="https://drive.google.com/uc?export=view&id=1AyNS0vsdDXoJHDNIAtg8JvT28t5weSSH" style="height: 3rem"/>
</div>
<div align="center">
<em>Total measures per Market, Table from Tableau Dashboard</em>
</div>
 </p>


Time series Global stores has experienced growth in sales and profits gradually from 2011 to 2014. If this trend continues, the future is bright for the business.


 <p>
<div align="center">
<img src="https://drive.google.com/uc?export=view&id=1B-niEIhq4lGEipus60HO4qjivusBhs-E" style="height: 3rem"/>
</div>
<div align="center">
<em>Yearly Profit Forecasted Line Chart from Tableau Dashboard</em>
</div>
 </p>



* Limitations

The data lacks population numbers for the different regions. Demographic information would have let us know markets where Global store is most popular and markets where more advertising and marketing needs to take place. Markets with high populations are likely to record high sales by default. This however doesn't represent the actual market share in comparison with key competitors.

* Recommendations

From the analysis, the tables sub-category is making losses. Only 20 out of 170 table types generate a profit margin that is higher than the company's average profit margin. Global stores can continue selling the 20 profitable unique tables. On the other hand, the loss making tables can be gradually removed from the online shelfs for higher profitability.

## Data Analytics using Rill

Rill is the fastest path from data lake to dashboard. Download Rill to start modeling data and create fast, exploratory dashboards.


 <p>
<div align="left">
<a href="https://www.rilldata.com/" target="_blank"> <img src="https://drive.google.com/uc?export=view&id=1AhbP-OjxxSEuFll_GH44burZnqh9V8ln" style="height: 3rem"/> </a>
</div>
 </p>


According to its official site, [Rill](https://www.rilldata.com/) is the fastest path from data lake to dashboard. With Rill you can build radically simple, thoughtful, opinionated and, interactive dashboards as code. Each step from data to dashboard has versioning, Git sharing, and easy project rehydration.

Download Rill to start modeling data and create fast, exploratory dashboards. Here we will not talk about the details on Rill's installation, I wrote an [article](https://medium.com/data-engineer-things/unlocking-data-insights-with-rill-a-comprehensive-guide-to-streamlined-data-analytics-41d83e06966d) in Medium that explains in detail how you can configure and start modeling your data with Rill.

## Global Superstore dashboard built in Rill

The Dashboard consumes data from 3 CSV files made available publicly in an S3 bucket:

 <p>
<div align="center">
<img src="https://drive.google.com/uc?export=view&id=1B8gNBCl1sgkRqL9h_pilt0Sbfz4CZn6X" style="height: 3rem"/>
</div>
<div align="center">
<em>Data for Rill Dashboard stored in S3 bucket</em>
</div>
 </p>

The data sources consist of 3 tables, `orders`, `returns` and `people`, being Orders the table where we will perform the majority of our analysis.

In the Rill Dashboard, we have focused the analysis on the `Sales`, `Profit` and `Quantities` sold and number of Orders processed over the time dimension of the dataset, that corresponds to the `Order Date` field.

This is precisely where Rill shines, it models your numerical variables (`Sales`, `Profit` `Quantities`), and aggregates your categorical data (`Customers`,`Country`,`Markets`,`Region`,`Segment`) over a DateTime variable (`Order Date`).

The Rill Dashboard was already deployed and is available in the [UI](https://ui.rilldata.com/fvgm-spec/global_superstore/model_dashboard). As seen in the image below shows how Rill organizes the data based on the definitions set in [YAML files](https://github.com/fvgm-spec/global_superstore/blob/main/dashboards/model_dashboard.yaml)


 <p>
<div align="center">
<a href="https://ui.rilldata.com/fvgm-spec/global_superstore/model_dashboard" target="_blank"><img src="https://drive.google.com/uc?export=view&id=1AchaybV_ji0EqOlyrVbb2z_mC_XndBmI" style="height: 3rem"/> </a>
</div>
<div align="center">
<em>Rill Dashboard UI</em>
</div>
 </p>


The Measures (Numerical variables) are visualized over the range of the DateTime dimension, and these measures can be filtered by the Categorical dimensions.




## Bonus track


We can be able to perform a similar data analysis as the one we did using Tableau and Rill, just opening the dataset as a [Google Sheet](https://docs.google.com/spreadsheets/d/1zs9qS-9Tb23dyhRzvek8nj3bKMvQyc9Ft6dq3oPqfSI/edit#gid=2124101240), and then querying the data of the `Orders` dataset with the Google Sheets [QUERY function](https://support.google.com/docs/answer/3093343?hl=en).

 <p>
<div align="center">
<a href="https://docs.google.com/spreadsheets/d/1zs9qS-9Tb23dyhRzvek8nj3bKMvQyc9Ft6dq3oPqfSI/edit#gid=2124101240" target="_blank"><img src="https://drive.google.com/uc?export=view&id=1BAtfnJZWkO1xyW2uLRmOh9Umol5SV9lb" style="height: 3rem"/> </a>
</div>
<div align="center">
<em>Google Sheets Query function</em>
</div>
 </p>
